pydantic/types.py

```diff
@@ -237,4 +240,26 @@ class NegativeFloat(ConstrainedFloat):
     lt = 0
 
 
+class ConstrainedUUID(UUID):
+    version_: int = None
+
+    @classmethod
+    def get_validators(cls):
+        yield uuid_validator
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value: UUID) -> UUID:
+        if cls.version_ and value.version != cls.version_:
+            raise ValueError(f'uuid version {cls.version_} expected, not {value.version}')
+
+        return value
+
+
+def conuuid(*, version=4) -> Type[UUID]:
+    # use kwargs then define conf in a dict to aid with IDE type hinting
+    namespace = dict(version_=version)
+    return type('ConstrainedUUIDValue', (ConstrainedUUID,), namespace)
```

Comment:
Given that there are only 4 times, should it would be cleaner to just have 4 tyes: `Uuid1`, `Uuid2`, ...

Rather than having to implement `conuuid`

Reply:
You're right!